File: backend/common/src/main/scala/no/ndla/common/model/domain/draft/Grade.scala
/*
 * Part of NDLA common
 * Copyright (C) 2024 NDLA
 *
 * See LICENSE
 *
 */

package no.ndla.common.model.domain.draft

import enumeratum.values.{IntCirceEnum, IntEnum, IntEnumEntry}
import sttp.tapir.Schema
import sttp.tapir.codec.enumeratum.*

sealed abstract class Grade(val value: Int) extends IntEnumEntry

object Grade extends IntEnum[Grade] with IntCirceEnum[Grade] {

  val values: IndexedSeq[Grade]      = findValues
  implicit val schema: Schema[Grade] = schemaForIntEnumEntry[Grade]

  case object One   extends Grade(1)
  case object Two   extends Grade(2)
  case object Three extends Grade(3)
  case object Four  extends Grade(4)
  case object Five  extends Grade(5)
}

File: backend/taxonomy-api/src/main/kotlin/no/ndla/taxonomy/domain/Grade.kt
/*
 * Part of NDLA taxonomy-api
 * Copyright (C) 2024 NDLA
 *
 * See LICENSE
 */

package no.ndla.taxonomy.domain

import com.fasterxml.jackson.annotation.JsonValue
import io.swagger.v3.oas.annotations.media.Schema

@Schema(enumAsRef = true)
enum class Grade(val value: Int) {
  One(1),
  Two(2),
  Three(3),
  Four(4),
  Five(5),
  ;

  @JsonValue fun toInt() = value

  companion object {
    fun fromInt(value: Int): Grade =
        when (value) {
          1 -> One
          2 -> Two
          3 -> Three
          4 -> Four
          5 -> Five
          else -> throw IllegalArgumentException("Unexpected grade value: $value. Must be 1-5.")
        }
  }
}
